torch/distributed/tensor/parallel/ddp.py

```diff
@@ -0,0 +1,93 @@
+from typing import Any, List, Tuple
+
+import torch.nn as nn
+from torch.distributed.tensor.parallel.fsdp import _flatten_tensor, _unflatten_tensor
+
+__all__ = ["pre_dp_model_transform"]
+
+
+def _get_submodule_n_params(module: nn.Module, path: str):
+    """
+    Get submodule and the direct path of parameter from the module
+    """
+    if "." in path:
+        path_list = path.split(".")
+        parent_module_path = ".".join(path_list[:-1])
+        module = module.get_submodule(parent_module_path)
+        path = path_list[-1]
+    return module, path
+
+
+def _update_model_param(param_list: List[Tuple[nn.Module, str, nn.Parameter]]):
+    """
+    Update parameters within the model
+    """
+    for item in param_list:
+        parent_module, module_path, t = item
+        assert hasattr(parent_module, module_path)
+        delattr(parent_module, module_path)
+        setattr(parent_module, module_path, t)
+
+
+def _reconstruct_dtensor(model: nn.Module, _input: Any):
+    """
+    Recontruct DTensor parameters from local tensors
+    """
+    param_list = []
+    for name, t in model.named_parameters():
+        if hasattr(t, "_st_info"):
+            dtensor = _unflatten_tensor(t, t._st_info)
+            param_list.append((*_get_submodule_n_params(model, name), dtensor))
+    _update_model_param(param_list)  # type: ignore[arg-type]
+
+
+def _localize_dtensor(model: nn.Module, _input: Any, _output: Any):
+    """
+    Convert DTensor parameters to local tensors
+    """
+    param_list = []
+    for name, param in model.named_parameters():
+        t, sharding_info = _flatten_tensor(param)
+        if sharding_info is not None:
+            t = nn.Parameter(t)
+            t._st_info = sharding_info  # type: ignore[attr-defined]
+            param_list.append((*_get_submodule_n_params(model, name), t))
+    _update_model_param(param_list)  # type: ignore[arg-type]
+
+
+def pre_dp_model_transform(model: nn.Module):
```

Comment:
I recommend naming the arg `module` instead of `model` as we push for distributed API composability since `model` generally refers to the root module, whereas `module` could mean a submodule.

In that case, you may also prefer the function name `pre_dp_module_transform`.

Reply:
More broadly, I was wondering: Should we build this logic into DDP/FSDP and hide it from the user (avoiding the extra call)?

The natural follow-up question is how would users disable this logic if they are using DTensor in their own way and do not want this conversion logic done for them? In that case, our "official" TP API `parallelize_module()` can mark the constructed DTensors specially, and DDP/FSDP can only register this special logic if it detects such marked DTensors in their managed parameters.

What do you think?

Reply:
For FSDP, we already have the extension embedded inside FSDP already right? But this causes issues in state_dict so @fegin and @kumpera we think this would be a better UX for TP + DP.

Reply:
Also if users want to special handling we can either give user an option to register a customized handler here or user can choose not to call this API rather than embed into DDP/FSDP code?

Reply:
Actually, I also prefer to make this logic in FSDP and DDP. What I suggested is to use hook to implement which is what this PR does. But I think the information should be a state of DDP instead of using `_st_info` and is attached to the parameter.